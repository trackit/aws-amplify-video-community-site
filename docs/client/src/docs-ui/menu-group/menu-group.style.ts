import {css} from "emotion";

export const menuGroupHeaderStyle = css`
  width: 100%;
  text-align: initial;
  padding: 0.25rem 0.5rem;
  cursor: pointer;
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  appearance: none;
  background-color: transparent;

  h4 {
    cursor: pointer;
  }

  &:hover {
    background-color: var(--bg-color-hover);
  }

  &:active {
    color: var(--font-color);
  }
`;

export const arrowStyle = css`
  border: solid black;
  border-width: 0 0.125rem 0.125rem 0;
  padding: 0.1875rem;
  transition: all 0.25s cubic-bezier(0.165, 0.84, 0.44, 1);
`;

export const arrowDownStyle = css`
  margin-bottom: 0.125rem;
  transform: rotate(45deg);
`;

export const arrowUpStyle = css`
  margin-top: 0.125rem;
  transform: rotate(-135deg);
`;

export const linksStyle = css`
  display: flex;
  flex-direction: column;
  margin-bottom: 0.5rem;
`;

export const menuGroupItemStyle = css`
  stencil-route-link,
  a {
    display: inline-block;
    width: 100%;
    height: 100%;
  }
  a {
    padding: 0.25rem 0.5rem;
    font-size: 0.875rem;
    color: var(--font-color-secondary);
    &:hover {
      background-color: var(--bg-color-hover);
    }
  }
`;

export const activeLinkStyle = css`
  background-color: var(--bg-color-hover);
`;
