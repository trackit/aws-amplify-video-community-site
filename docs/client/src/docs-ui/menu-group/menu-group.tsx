import {Component, Prop, Host, h, State, Watch} from "@stencil/core";
import {MenuGroup, PageLink} from "../../api";
import {
  linksStyle,
  menuGroupHeaderStyle,
  menuGroupItemStyle,
  activeLinkStyle,
  arrowStyle,
  arrowDownStyle,
  arrowUpStyle,
} from "./menu-group.style";
import {pageContext} from "../page/page.context";
import {SelectedFilters} from "../page/page.types";
import {sidebarLayoutContext} from "../../amplify-ui/sidebar-layout/sidebar-layout.context";
import {ToggleInView} from "../../amplify-ui/sidebar-layout/sidebar-layout.types";
import {MQTablet} from "../../amplify-ui/styles/media";

@Component({tag: "docs-menu-group", shadow: false})
export class DocsMenuGroup {
  /*** the group to render */
  @Prop() readonly menuGroup?: MenuGroup;
  /*** the filter key that applies to this product section */
  @Prop() readonly filterKey?: string;
  /*** the currently-selected filters */
  @Prop() readonly selectedFilters: SelectedFilters;
  /*** whether or not the menu is in view */
  @Prop() readonly inView?: boolean;
  /*** the menu toggle */
  @Prop() readonly toggleInView: ToggleInView;

  @State() expanded = false || this.filterKey === "integration";
  @State() itemsToDisplay?: PageLink[];

  toggleOpen = () => {
    this.expanded = !this.expanded;
  };

  shouldDisplay = ({filters}: PageLink): boolean => {
    return (
      // the filter key is undefined
      this.filterKey === undefined ||
      // this page is available independent of filter
      filters === undefined ||
      // this page is available in specific filtered versions (one of which is the globally-selected)
      (filters &&
        this.selectedFilters &&
        filters[this.filterKey].includes(
          this.selectedFilters[this.filterKey] as string,
        ))
    );
  };

  @Watch("menuGroup")
  @Watch("selectedFilters")
  componentWillLoad() {
    this.itemsToDisplay = this.menuGroup?.items.filter(this.shouldDisplay);
    const currentRoute = location.pathname.split("/q/").shift() as string;
    if (
      this.itemsToDisplay &&
      this.itemsToDisplay.some(({route}) => route === currentRoute)
    ) {
      this.expanded = true;
    }
  }

  closeMenuIfOnMobile = () => {
    // Slice off the "@media " string at the start for use in JS instead of CSS
    const MQTabletJS = MQTablet.substring(6);
    const isMobile = !window.matchMedia(MQTabletJS).matches;
    if (isMobile && this.inView) {
      this.toggleInView();
    }
  };

  render() {
    return (
      <Host>
        {this.menuGroup &&
          this.itemsToDisplay &&
          this.itemsToDisplay.length > 0 && [
            <button class={menuGroupHeaderStyle} onClick={this.toggleOpen}>
              <h4>{this.menuGroup.title}</h4>
              <i
                class={{
                  [arrowStyle]: true,
                  [arrowUpStyle]: this.expanded,
                  [arrowDownStyle]: !this.expanded,
                }}
              ></i>
            </button>,
            this.expanded && (
              <div class={linksStyle}>
                {this.itemsToDisplay.map((item) => (
                  <docs-internal-link
                    key={item.title}
                    href={item.route}
                    class={menuGroupItemStyle}
                    activeClass={activeLinkStyle}
                    onClick={this.closeMenuIfOnMobile}
                  >
                    {item.title}
                  </docs-internal-link>
                ))}
              </div>
            ),
          ]}
      </Host>
    );
  }
}

pageContext.injectProps(DocsMenuGroup, ["selectedFilters"]);
sidebarLayoutContext.injectProps(DocsMenuGroup, ["toggleInView", "inView"]);
