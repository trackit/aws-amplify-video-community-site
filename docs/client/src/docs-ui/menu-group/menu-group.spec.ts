import {DocsMenuGroup} from "./menu-group";
import {newSpecPage} from "@stencil/core/testing";

describe("docs-menu-group", () => {
  describe("Component logic", () => {
    let menu: DocsMenuGroup;
    beforeEach(() => (menu = new DocsMenuGroup()));

    it("should init `menuGroup` as `undefined`", () =>
      expect(menu.menuGroup).toBeUndefined());
  });

  describe("Render logic", () => {
    it("should render", async () => {
      expect(
        (
          await newSpecPage({
            components: [DocsMenuGroup],
            html: `<docs-menu-group />`,
          })
        ).root,
      ).toMatchSnapshot();
    });
  });
});
