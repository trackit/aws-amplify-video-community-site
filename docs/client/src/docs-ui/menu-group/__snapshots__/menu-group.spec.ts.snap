// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`docs-menu-group Render logic should render 1`] = `<docs-menu-group></docs-menu-group>`;
